feat: enable uninstalling PatchedMediaKeysApple

martinstark opened the pull request:
Close #4469

The idea is to only enable uninstalling when it is desired, in order to not hold things in memory unnecessarily.

martinstark commented:
Will re-open once I run some more tests.

martinstark commented:
Re-opening to ask questions

martinstark commented on lib/polyfill/patchedmediakeys_apple.js:
```diff
@@ -27,14 +27,29 @@ goog.require('shaka.util.StringUtils');
+    if (enableUninstall) {
+      this.enableUninstall = true;
+      // eslint-disable-next-line no-restricted-syntax
+      this.originalHTMLMediaElementPrototypeMediaKeys = HTMLMediaElement
```
This does not work (cannot call get directly on prototype.mediaKeys), and I'm unsure if there's a need for saving and restoring the mediaKeys. Any input would be appreciatedb

joeyparrish replied:
If you need to backup and restore a getter, you may need to try getOwnPropertyDescriptor.  There is a good example of backing up and restoring a getter this way in test/player_unit.js.

Does this help?

joeyparrish commented:
I don't understand this PR or the associated issue, and it seems really odd to me.  Let's discuss on the issue, but for now, please do not merge the PR.  Sorry for the inconvenience.

joeyparrish commented on lib/polyfill/patchedmediakeys_apple.js:
```diff
@@ -59,6 +74,38 @@ shaka.polyfill.PatchedMediaKeysApple = class {
+   * Uninstalls the polyfill if needed and enabled.
+   * @export
+   */
+  static uninstall() {
```
How difficult would it be to add a test that exercises this code?

martinstark commented:
Thanks @joeyparrish, I'll look at adjusting according to your feedback and add a test

martinstark commented:
@avelad @joeyparrish I've addressed the feedback, introduced unit tests and added a short blurb in the documentation. 

I opted for not restricting the unit test to only run on Safari. The functionality that's being tested is just overriding and restoring browser globals.

I don't know of a way of checking strict equality on the mediaKeys getter. The test does a best effort check to see if it looks like the value is no longer overridden, but it doesn't guarantee that it was restored.

I have not yet run the latest code updates on a Mac, I'll update the PR once I have.

avelad commented:
@martinstark It seems that the tests have failed, can you check it?

martinstark commented:
@avelad 
```bash
 1) cues ending exactly now
     TextDisplayer layout using browser-native rendering
     native-end-time-edge-case: Expected 0.9133498736570255 not to be less than 0.95.
```
I don't think this is related to my changes here, what do you think?

avelad commented:
I'll re-run the failed jobs

martinstark commented:
@avelad the safari test still fails, I've rebased on the lastest `main` to see if that makes a difference

martinstark commented:
```bash
1) cues ending exactly now
     TextDisplayer layout using browser-native rendering
     native-end-time-edge-case: Expected 0.9133498736570255 not to be less than 0.95.
```
Keeps failing with exactly the same number. I've had a brief look at the test in question, as well as the text displayer, but no ideas spring to mind.

avelad commented:
I'll ignore the fail test to approve the PR. Please send the final code please (no comments in the tests)!

martinstark commented:
@avelad I tried to remove the unit tests to see if that's what was causing the Safari test suite failure. I've reverted this and squashed everything.

Thanks for your time!

avelad commented on docs/tutorials/fairplay.md:
````diff
@@ -8,6 +8,13 @@ Modern EME yet, you can use legacy Apple Media Keys with:
+If you need to use both legacy and Modern EME, for example if you have to support
+multiple DRM providers, it is possible to enable uninstalling the polyfill:
+```js
+shaka.polyfill.PatchedMediaKeysApple.install(/* enableUninstall */ true);
````
This should be: `/* enableUninstall= */ true`. This keeps it consistent with the rest of the code.

martinstark replied:
addressed

avelad reviewed (changes requested):
A minor nit, otherwise LGTM

martinstark commented:
@avelad not sure what your process is, I just squashed and force-pushed after you approval, but that dismisses your review. Any input?

avelad commented:
The squash it's done by github, so really it's not a problem :) I'll approve again

avelad commented:
@joeyparrish can you review it? Thanks!
